daemon/containerd/image_list.go

```diff
@@ -66,63 +70,32 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 		root = make([]*[]digest.Digest, len(imgs))
 		layers = make(map[digest.Digest]int)
 	}
+
 	for n, img := range imgs {
 		if !filterFn(img) {
 			continue
 		}
 
-		diffIDs, err := img.RootFS(ctx)
+		platforms, err := images.Platforms(ctx, i.client.ContentStore(), img.Target)
 		if err != nil {
 			return nil, err
 		}
-		chainIDs := identity.ChainIDs(diffIDs)
-		if opts.SharedSize {
-			root[n] = &chainIDs
-			for _, id := range chainIDs {
-				layers[id] = layers[id] + 1
-			}
-		}
 
-		size, err := img.Size(ctx)
-		if err != nil {
-			return nil, err
-		}
+		for m, platform := range platforms {
+			image, chainIDs, err := i.singlePlatformImage(ctx, img, platform, opts)
+			if err != nil {
+				return nil, err
+			}
```

Comment:
`image`can be nil without `err != nil` (if some content for that platform is missing)

```suggestion
			if err != nil {
				return nil, err
			}
			if image == nil {
				continue
			}
```